arwen/elrondapi/elrondei.go

```diff
@@ -627,3 +656,280 @@ func int64finish(context unsafe.Pointer, value int64) {
+	gasToUse += ethContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	ethContext.UseGas(gasToUse)
+
+	if ethContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
+	bigIntVal := big.NewInt(0).SetBytes(value)
+	ethContext.Transfer(dest, send, bigIntVal, nil)
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  send,
+			Arguments:   data,
+			CallValue:   bigIntVal,
+			GasPrice:    0,
+			GasProvided: uint64(gasLimit),
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+	err := ethContext.ExecuteOnDestContext(contractCallInput)
+	if err != nil {
+		return 1
+	}
+
+	return 0
+}
+
+//export executeOnDestContext
+func executeOnDestContext(
+	context unsafe.Pointer,
+	gasLimit int64,
+	addressOffset int32,
+	valueOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	ethContext := arwen.GetEthContext(instCtx.Data())
+
+	send := ethContext.GetSCAddress()
+	dest := arwen.LoadBytes(instCtx.Memory(), addressOffset, arwen.AddressLen)
+	value := arwen.LoadBytes(instCtx.Memory(), valueOffset, arwen.BalanceLen)
+	function, data, actualLen := getArgumentsFromMemory(context, functionOffset, functionLength, numArguments, argumentsLengthOffset, dataOffset)
+
+	gasToUse := ethContext.GasSchedule().ElrondAPICost.ExecuteOnDestContext
+	gasToUse += ethContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	ethContext.UseGas(gasToUse)
+
+	if ethContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
+	ethContext.Transfer(dest, send, big.NewInt(0).SetBytes(value), nil)
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  send,
+			Arguments:   data,
+			CallValue:   big.NewInt(0).SetBytes(value),
+			GasPrice:    0,
+			GasProvided: uint64(gasLimit),
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+	err := ethContext.ExecuteOnSameContext(contractCallInput)
+	if err != nil {
+		return 1
+	}
+
+	return 0
+}
+
+func getArgumentsFromMemory(
+	context unsafe.Pointer,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (string, [][]byte, int32) {
+	instCtx := wasmer.IntoInstanceContext(context)
+	argumentsLengthData := arwen.LoadBytes(instCtx.Memory(), argumentsLengthOffset, numArguments*4)
+
+	currOffset := dataOffset
+	data := make([][]byte, numArguments)
+	for i := int32(0); i < numArguments; i++ {
+		currArgLenData := argumentsLengthData[i*4 : i*4+4]
+		actualLen := dataToInt32(currArgLenData)
+
+		data[i] = arwen.LoadBytes(instCtx.Memory(), currOffset, actualLen)
+		currOffset += actualLen
+	}
+
+	function := arwen.LoadBytes(instCtx.Memory(), functionOffset, functionLength)
+
+	return string(function), data, currOffset - dataOffset
+}
+
+//export delegateExecution
+func delegateExecution(
```

Comment:
This function is almost identical to `executeOnSameContext()`. Could they have their inner functionality merged into a separate function, which is then called differently in the bodies of `executeOnSameContext()` and `delegateExecution()`?

Reply:
Same with `executeReadOnly()` as well.